theano/compile/builders.py

```diff
@@ -1,44 +1,121 @@
+"""Define new Ops from existing Ops"""
 from __future__ import absolute_import, print_function, division
+from functools import reduce, partial
+from collections import OrderedDict
+
 import theano
 from theano import gof
 from theano.compat import izip
 from theano.compile.function_module import orig_function
-from theano.compile import SharedVariable, rebuild_collect_shared
+from theano.compile import SharedVariable, rebuild_collect_shared, optdb
 from theano.gof import ops_with_inner_function
 from theano.gof.graph import io_connection_pattern
-
-from functools import reduce
+from theano.gof.null_type import NullType
+from theano.gradient import DisconnectedType
 
 
 class OpFromGraph(gof.Op):
     """
-    This creates an `Op` from inputs and outputs lists of variables.
-
-    The signature is similar to theano.function() and the resulting
-    `Op`'s perform will do the same operation as::
+    This creates an ``Op`` from inputs and outputs lists of variables.
+    The signature is similar to :func:`theano.function <theano.function>`
+    and the resulting ``Op``'s perform will do the same operation as::
 
         orig_function(inputs, outputs, **kwargs)
 
-    TODO:
+    Currently does not support ``updates`` or ``givens`` argument.
+
+    Parameters
+    ----------
+
+    inputs: list of :class:`Variable <theano.gof.Variable>`
+
+    outputs: list of :class:`Variable <theano.gof.Variable>`
+
+    inline: bool, optional
+        Defaults to ``False``
+
+        ``True`` : Cause the Op's original graph being used during
+        compilation, the Op will not be visible in the compiled
+        graph but rather its internal graph.
+
+        ``False`` : will use a pre-compiled function inside.
+
+    grad_overrides : single or list of {0, None, Ellipsis, OpFromGraph, callable}, optional
+        Defaults to ``None``.
+
+        ``None`` : No value, gives NullType()
+        ``0`` : zero value, gives DisconnectedType()
+        ``...`` : Do not override, use default grad() result
+
+        OpFromGraph instance : Override with another OpFromGraph, should
+        accept inputs as the same order and types of "inputs" and "output_grads"
+        arguments as one would specify in grad() method.
+
+        callable : similar to OpFromGraph instance, must return list of
+        :class:`Variable <theano.gof.Variable>`.
+
+        list: Each OpFromGraph/callable must return a single
+        :class:`Variable <theano.gof.Variable>`. Each list element corresponds to gradient of
+        a specific input, length of list must be equal to number of inputs.
+
+    rop_overrides : single or list of {0, None, Ellipsis, OpFromGraph, callable}, optional
+        Defaults to ``None``.
+
+        ``None`` : No value, gives NullType()
+        ``0`` : zero value, gives zeros_like(...)
+        ``...`` : Do not override, use default R_op() result
```

Comment:
This is a tough interface question, and I'm not sure how to solve it, but here's my take.
- I would like to avoid the use of `None`, since we used to have it mean an ambiguous "not defined or not connected or zero" in the past for `grad`, and it is still the case for `R_Op`.
- Following "explicit is better than implicit" from the _Zen of Python_, I would say the way of specifying `NullType()` or `DisconnectedType()` should be to pass it (maybe accept either the class or an instance, so user can define their own custom message, but do not have to). The option to return `DisconnectedType()` seems missing in the current version, btw.
- The use of `Ellipsis` is clever, and I like its compactness. I'm afraid that it may be _too_ clever, or not clear enough, or that examples look like incomplete code. However, the only alternative I have for that would be `False`, as in "do not override".

Let's discuss that and agree on a solution before implementing it, as it is going to be hard to change after the PR is merged.